BFD-3091: Use dynamic versions in ansible

mjburling opened the pull request:
**JIRA Ticket:**
[BFD-3091](https://jira.cms.gov/browse/BFD-3091)

**User Story or Bug Summary:**
Replace static, 1.0.0-SNAPSHOT references with a dynamic version in workflows centered around ansible.

---

### What Does This PR Do?
This introduces 2 larger changes across BFD's delivery and deployment mechanisms:
* standardize the ansible test _harness_ and scripting entrypoints to accept an optional image flag and mandatory `bfd_version` positional argument for ansible roles supporting bfd-db-migrator, bfd-pipeline, and bfd-server.
* generally defer to the version encoded in `apps/pom.xml` as a stored, ansible variable `bfd_version` _or_ accurately divine script/artifact locations through other, automatic means where possible

#### Future Work
* Inputs for `bfd_version` and the ansible default value is 1.0.0-SNAPSHOT. It might be better to take this default away on the heels of BFD-2906, et al
* The `bfd_version` is a bit of a kludge to facilitate testing the ansible roles–there are some fairly straightforward ways to avoid this, it will just take a little more time, effort, and a few additional lines to accommodate. Our role testing executes both install and launch _scopes_ (installation (via `--tag pre-ami`) is where a version string can be helpful) which makes a `bfd_version` variable too convenient ☹️
* `bfd_version` would probably better serve us as an override; in the absence of an override, the automation should take care of itself to find the appropriate artifacts and paths
* @billie-rose pointed out some of the weirdness surrounding the resultant `ansible-playbook` call that tends to result in a functional, if marked, doubling of `--extra-vars` without arguments, e.g. `ansible-playbook test_basic.yml --inventory=inventory.docker.yaml --extra-vars --extra-vars bfd_version="1.0.0-SNAPSHOT"`. Along with the other treatment of `bfd_version` (further evidence that we need to chase a better solution), we should address this, too.


### What Should Reviewers Watch For?

* Verify successful deployment to `test`: https://jenkins-east.cloud.cms.gov/bfd/job/bfd-build-deploy/job/PR-2093/ 
* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @<!-- -->StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A


### Submitter Checklist
<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

mackec1 commented on apps/build.groovy:
```diff
@@ -94,10 +94,10 @@ EOF
-		dataPipelineZip: 'apps/bfd-pipeline/bfd-pipeline-app/target/bfd-pipeline-app-1.0.0-SNAPSHOT.zip',
-		dataServerLauncher: 'apps/bfd-server/bfd-server-launcher/target/bfd-server-launcher-1.0.0-SNAPSHOT.zip',
-		dataServerWar: 'apps/bfd-server/bfd-server-war/target/bfd-server-war-1.0.0-SNAPSHOT.war'
+		dbMigratorZip:      sh(returnStdout: true, script: """find "${workspace}/apps/bfd-db-migrator/target" -type f -name bfd-db-migrator-*.zip""").trim(),
```
just curious why you searched for both 'type' and then 'name'...

mjburling replied:
It's _probably_ not necessary here. At worst, the inclusion of `-type f` is harmless, but it is more explicit and potentially safer to limit the result set to files rather than say... directories?

There isn't a lot of safety in this process, but if a directory somehow came to be named e.g. `bfd-db-migrator-1.0.0-SNAPSHOT.zip` here and was filtered out as a consequence of the `type -f`, the packer build would fail due to an inaccessible _or_ non-existent path on the ansible host (at Jenkins) represented by an empty `db_migrator_zip` ansible variable. If a _directory_ were passed in without this filtering, packer+ansible build would receive this mysterious directory, but it might be less obvious and take longer to reach failure. The problem would be a little later on the EC2 instance targeted by packer+ansbile. Hmm.

mackec1 commented on ops/ansible/roles/bfd-pipeline/README.md:
```diff
@@ -20,7 +20,7 @@ Role Variables
 This role is highly configurable, though it tries to provide reasonable defaults where possible. Here are the variables that must be defined by users:
 
-    data_pipeline_zip: /home/karlmdavis/workspaces/cms/beneficiary-fhir-data.git/apps/bfd-pipeline/bfd-pipeline-app/target/bfd-pipeline-app-1.0.0-SNAPSHOT.zip
+    data_pipeline_zip: /home/karlmdavis/workspaces/cms/beneficiary-fhir-data.git/apps/bfd-pipeline/bfd-pipeline-app/target/bfd-pipeline-app-2.0.0-SNAPSHOT.zip
```
karl lives on forever???

mackec1 reviewed (approved):
looks good

brandoncruz3 reviewed (approved):
Nice!

billie-rose commented:
Will the ansible-playbook `--extra-vars` command work correctly if `$EXTRA_VARS` is empty? I noticed we added it to some of the commands where it was previously missing. I think running `ansible --extra-vars --extra-vars "foo=bar" would result in `ansible: error: argument -e/--extra-vars: expected one argument`

mjburling commented:
> Will the ansible-playbook `--extra-vars` command work correctly if `$EXTRA_VARS` is empty? I noticed we added it to some of the commands where it was previously missing. I think running `ansible --extra-vars --extra-vars "foo=bar"` would result in `ansible: error: argument -e/--extra-vars: expected one argument`

That's a good observation, but the passing tests here suggest that, at least with this version of ansible, an `--extra-vars --extra-vars "foo=bar"` is legal. Admittedly, I didn't spend all that long implementing any of this yesterday, and it worked and the crumby ansible tests we have here continue to work.

Let me meditate on whether we want something more thoughtful on this whole thing that's providing... _some_ value for at least a few more weeks. 🤔

mjburling commented:
> > Will the ansible-playbook `--extra-vars` command work correctly if `$EXTRA_VARS` is empty? I noticed we added it to some of the commands where it was previously missing. I think running `ansible --extra-vars --extra-vars "foo=bar"` would result in `ansible: error: argument -e/--extra-vars: expected one argument`
> 
> That's a good observation, but the passing tests here suggest that, at least with this version of ansible, an `--extra-vars --extra-vars "foo=bar"` is legal. Admittedly, I didn't spend all that long implementing any of this yesterday, and it worked and the crumby ansible tests we have here continue to work.
> 
> Let me meditate on whether we want something more thoughtful on this whole thing that's providing... _some_ value for at least a few more weeks. 🤔

@billie-rose I've added another bullet under future work here. I think this will be a _fairly_ temporary solution as we work through the final details of enabling the maven release plugin under BFD-2906 this week. 🤞
